packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -347,80 +341,55 @@ export class PendingStateManager implements IDisposable {
         this.clientId = this.stateHandler.clientId();
 
-        assert(this.initialStates.isEmpty(), 0x174 /* "initial states should be empty before replaying pending" */);
+        assert(this.initialMessages.isEmpty(), 0x174 /* "initial states should be empty before replaying pending" */);
 
-        let pendingStatesCount = this.pendingStates.length;
-        if (pendingStatesCount === 0) {
+        let pendingMessagesCount = this.pendingMessages.length;
+        if (pendingMessagesCount === 0) {
             return;
         }
 
-        // Reset the pending message count because all these messages will be removed from the queue.
-        this._pendingMessagesCount = 0;
-
-        const messageBatchQueue = new Deque<IPendingMessage>();
-
-        // Process exactly `pendingStatesCount` items in the queue as it represents the number of states that were
+        // Process exactly `pendingMessagesCount` items in the queue as it represents the number of messages that were
         // pending when we connected. This is important because the `reSubmitFn` might add more items in the queue
         // which must not be replayed.
-        while (pendingStatesCount > 0) {
+        while (pendingMessagesCount > 0) {
             // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-            const pendingState = this.pendingStates.shift()!;
-            switch (pendingState.type) {
-                case "message":
-                    assert(pendingState.opMetadata?.batch !== false || messageBatchQueue.length > 0,
-                        0x41b /* We cannot process batches in chunks */);
-                    /**
-                     * We want to ensure grouped messages get processed in a batch.
-                     * Note: It is not possible for the PendingStateManager to receive a partially acked batch. It will
-                     * either receive the whole batch ack or nothing at all.
-                     */
-                    if (messageBatchQueue.length > 0 || pendingState.opMetadata?.batch) {
-                        messageBatchQueue.enqueue(pendingState);
-                    } else {
+            let pendingMessage = this.pendingMessages.shift()!;
+            pendingMessagesCount--;
+            assert(pendingMessage.opMetadata?.batch !== false, 0x41b /* We cannot process batches in chunks */);
+
+            /**
+             * We want to ensure grouped messages get processed in a batch.
+             * Note: It is not possible for the PendingStateManager to receive a partially acked batch. It will
+             * either receive the whole batch ack or nothing at all.
+             */
+            if (pendingMessage.opMetadata?.batch) {
+                assert(pendingMessagesCount > 0, "Last pending message cannot be a batch begin");
+
+                this.stateHandler.orderSequentially(() => {
+                    while (pendingMessagesCount >= 0) { // check is >= because batch end may be last pending message
```

Comment:
There should probably be validation inside this loop that we do infact get a batch end. Otherwise, if this happens it would be very hard to debug.